security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,107 @@
+package pl.touk.nussknacker.ui.security.api
+
+import akka.http.scaladsl.server.{AuthorizationFailedRejection, Directive0, Directive1, RejectionHandler}
+import akka.http.scaladsl.server.Directives.{handleRejections, optionalHeaderValueByName, provide, reject}
+import cats.data.OptionT
+import pl.touk.nussknacker.security.AuthCredentials.{
+  ImpersonatedAuthCredentials,
+  NoCredentialsProvided,
+  PassedAuthCredentials
+}
+import pl.touk.nussknacker.security.{AuthCredentials, ImpersonatedUserIdentity}
+import pl.touk.nussknacker.ui.security.api.AuthenticationManager.{
+  impersonateHeaderName,
+  impersonationHeaderEndpointInput,
+  mappedAuthenticationEndpointInput
+}
+import sttp.tapir._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuthenticationManager(authenticationResources: AuthenticationResources)(implicit ec: ExecutionContext) {
+
+  lazy val authenticationRules: List[AuthenticationConfiguration.ConfigRule] =
+    authenticationResources.configuration.rules
+  lazy val isAdminImpersonationPossible: Boolean = authenticationResources.configuration.isAdminImpersonationPossible
+
+  def authenticate(): Directive1[AuthenticatedUser] = authenticationResources.getAnonymousRole match {
+    case Some(role) =>
+      authenticationResources.authenticate().optional.flatMap {
+        case Some(user) => provideOrImpersonateUser(user)
+        case None => handleAuthorizationFailedRejection.tmap(_ => AuthenticatedUser.createAnonymousUser(Set(role)))
+      }
+    case None => authenticationResources.authenticate().flatMap(provideOrImpersonateUser)
+  }
+
+  def authenticate(authCredentials: AuthCredentials): Future[Option[AuthenticatedUser]] = authCredentials match {
+    case passedCredentials @ PassedAuthCredentials(_) => authenticationResources.authenticate(passedCredentials)
+    case NoCredentialsProvided                        => authenticateWithAnonymousAccess()
+    case ImpersonatedAuthCredentials(impersonatingUserAuthCredentials, impersonatedUserIdentity) =>
+      authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserIdentity)
+  }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] =
+    authenticationResources
+      .authenticationMethod()
+      .and(impersonationHeaderEndpointInput)
+      .map(mappedAuthenticationEndpointInput)
+
+  private def provideOrImpersonateUser(user: AuthenticatedUser): Directive1[AuthenticatedUser] = {
+    optionalHeaderValueByName(impersonateHeaderName).flatMap {
+      case Some(impersonatedUserIdentity) =>
+        authenticationResources.getImpersonatedUserData(impersonatedUserIdentity) match {
+          case Some(impersonatedUserData) =>
+            provide(AuthenticatedUser.createImpersonatedUser(user, impersonatedUserData))
+          case None => reject
```

Comment:
I've previously added a comment about more symmetric implementation. I'm not sure but here it looks like it behaves a little bit differently than in the Tapir's case (the `None`s, in Tapir's case, are flattened, so it looks like it doesn't behave exactly the same). Maybe it'll be easier to see it when you do the changes mentioned in my previous comment.

Reply:
As I mentioned it doesn't seem trivial to make it more symmetric. 
I looked into the logic and in case of Tapir I did not differenate between failure in impersonating user's authentication and failure in fetching impersonated user's data. I modified the code to operate on `Either` in case of Tapir authentication to handle that situation. I moved `SecurityError` classes to the security module. It makes more sense for them to be there and it made it easier to handle error responses for both Tapir and Akka.